client/src/logic/impl/MapNavigatorImpl.ts

```diff
@@ -1,19 +1,92 @@
+import { Constants } from "../../utils/Constants";
 
 export class MapNavigatorImpl implements MapNavigator {
-    
+    private graph: Graph;
+
+    constructor(graph: Graph) {
+        this.graph = graph;
+    }
+
     findShortestPath(startNodeId: number, endNodeFilter: MapNodeFilter): NavigationDirections {
         let path = this.findPathToNearestNode(startNodeId, endNodeFilter);
         let navNodes = this.convertPathToNavigationNodes(path);
         return this.splitNavigationPathIntoSteps(navNodes);
     }
 
-    private findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
-        throw new Error("Method not implemented.");
+    /*private*/ findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
+
+        if (typeof this.graph.getNode(startNodeId) === "undefined") {
+            throw new Error("Start location not valid.");
+        }
+
+        let path: MapNode[] = [];
+        let neighbours: NeighbourConnection[];
+
+        let parents = new Map<number, number>(); //childId, parentId
+        let bestDistances = new Map<number, number>(); //nodeId, distance
+
+        let SortedSet = require('sorted-set');
+        let unvisitedNodes = new SortedSet({
+            hash(item: any) {return item.nodeId},
+            compare(a: any, b: any) {return a.distance - b.distance;}
+        });
+
+        let currentDistance: number;
+        let currentNodeId: number;
+        let allVisited: boolean;
+
+        bestDistances.set(startNodeId, 0);
+        unvisitedNodes.add({nodeId: startNodeId, distance: 0});
+
+        while (true) {
+            currentDistance = Constants.infinity;
+            allVisited = true;
+
+            if (unvisitedNodes.length > 0) {
+                currentNodeId = unvisitedNodes.head().nodeId;
+                currentDistance = unvisitedNodes.head().distance;
+                allVisited = false;
+            }
+
+            if (endNodeFilter.satisfies(this.graph.getNode(currentNodeId))) {
+                path.push(this.graph.getNode(currentNodeId));
+                
+                while (parents.has(currentNodeId)) {
+                    let parentId = parents.get(currentNodeId);
+
+                    path.push(this.graph.getNode(parentId));
+                    currentNodeId = parentId;
+                }
+
+                return path.reverse();
+
+            } else if (allVisited) {
+                throw new Error("Destination cannot be reached.");
+            }
+
+            unvisitedNodes.shift();
+            neighbours = this.graph.getNeighbours(currentNodeId);
+
+            for (let element of neighbours) {
+                let neighbourId = element.neighbour.id;
+                let neighbourDistance = element.distance;
+                
+                if (currentDistance + neighbourDistance < bestDistances.get(neighbourId)
+                    || !bestDistances.has(neighbourId)) {
+
```

Comment:
You should first check if bestDistances has this negithbour and then get it's value. This will result in unexpected behaviour if bestDistances does not have this neighbour.

The code should be as follows:
```ts
if (!bestDistances.has(neighbourId) || currentDistance + neighbourDistance < bestDistances.get(neighbourId)) {
```

If bestDistances does not have the neighbour it will return true and stop checking the rest of the if condition.

Reply:
Actually it doesn't result in an unexpected behaviour. It just returns false.
I wrote it this way because it seemed easier to read, but I will reverse them now.

Reply:
It works because of javascript (and typescript) specifics. If bestDistances map doesn't have a key neighbourId it doesn't know what to return (probably 0 or some default in this case). Then you compare that "null" result with numbers. In some other languages that would crash the execution.

Now after the fix, you first checked if the id is in the map and only after that do you read its value.

If you don't understand contact me informally. I'll explain it quickly.